### ISSUE TITLE
No types in comments?

### ISSUE THREAD

**Issue by trusktr** [microsoft/tsdoc#180]
I tried the playground, but it gives an error when I try something like

```
* @param {number} x - The first input number
```

---

The reason I'm asking is because I'd like to document my code that I migrated from JavaScript, and I don't want types to get in the way (like they do with TypeDoc).

I'm looking to write semantic documentation; the rest can be gleaned from intellisense so there's no need to write everything that intellisense covers in documentation.

What I have are "mixin classes", but I don't want the intricacy of the mixin implementation getting in the way of the documentation.

I'd like to document all my (mixin) class as a normal `@class`es, and make simple semantic documentation by hand (in jsdoc/tsdoc style comments).

For example, I'd like to do something like this:

```ts
import {Sizeable, TreeNode} from './'
import {Mixin, MixinResult, Constructor} from 'lowclass'

/**
 * @class Transformable
 * @mixes Sizeable
 * @mixes TreeNode
 */
function TransformableMixin<T extends Constructor>(Base: T) {

    // this "mixin class" mixes TreeNode and Sizeable
    class Transformable extends TreeNode.mixin(Sizeable.mixin(Constructor(Base))) {

        /**
         * Set the position of the Transformable.
         *
         * @property position
         * @memberof Transformable
         * @type {{x: number, y: number, z: number}}
         */
        set position(newValue: any) {
            this._setPropertyXYZ<Transformable, TransformProp>('position', newValue)
        }
        get position(): any {
            return this._props.position
        }

        // etc
    }

    return Transformable as MixinResult<typeof Transformable, T>
}

// this actually creates the `Transformable` class (with static .mixin helper).
export const Transformable = Mixin(TransformableMixin)
export interface Transformable extends InstanceType<typeof Transformable> {}
```

where `Sizeable` and `TreeNode` are "mixin classes" that are created also using the same pattern.

See what I'm trying to do?

So far, I face two major problems trying to write simple documentation like this:

1. No luck getting existing JSDoc generators to work with the TypeScript code: they all choke on the syntax or have some other unknown issues. I simply can't get them to work. They weren't designed with TypeScript in mind.
2. TypeScript doc generators (f.e. TypeDoc) seems to want to document all types under the sun exactly as technically correct. I don't want this, I just want simple documentation of my classes, and somewhere in the docs I will clearly state that all the classes are mixable (as an overall pattern across all classes). The output of TypeDoc is too technical, and prevents simple documentation when using non-regular patterns such as mixin classes as I am.

It'd be nice to have a JSDoc(TSDoc) parser that can ignore all my code, and just write the docs exactly as I've written them, while at the same time verifying that the types I've provided at least match correctly in most cases in the intellisense (or during tsc compile even, if that's possible).

No what I mean? I'd like to be the one in control of my docs, and not for the machine to be.

**Comment by trusktr** [comment]
Note, the `Transformable` "mixin class" can be used as either class, or a mixin:

```ts
import {Transformable} from './Transformable'

// instantiate a Transformable directly:
const f = new Transformable

// or extend from it:
class Foo extends Transformable {/*...*/}

// or mix it into a class
class Bar {/*...*/}
class Baz extends Transformable.mixin(Bar) {/*...*/}

// or mix it without extending
export default Transformable.mixin(Bar)
```

These mixin features are apparent in a bunch of my classes, but I want to document them in a simple way, not mechanically without control.

So I figure TSDoc might be the right tool to use, but I would like to have types in the documentation (even if just regular JSDoc-style types, at least something basic if it isn't actual TypeScript types (TS types may not be feasible the way I'm looking to document, and plus the same concept can apply across languages if I ever migrate)).

**Comment by trusktr** [comment]
TLDR, short version: The mixin features in my code are mostly for dev purposes, to compose features together, but the external use case is just to use the classes as custom elements, without thinking about mixin (and their types). I'd like to document them that way, as classes for the most part, to keep it simple. I don't necessarily need types from the code to propagate into the docs, but I do want the docs to work inside of TypeScript code (so I'm looking around for options).

**Comment by trusktr** [comment]
@octogonz @iclanton Hello guys, does tsdoc support type annotations? f.e. `@param {HERE} x - The first input number`?

**Comment by trusktr** [comment]
I deleted the irrelevant stuff I posted previously.

**Comment by octogonz** [comment]
In TypeScript, types are specified using in the language itself.  It doesn't make sense to declare them in the comment.  When people are writing new TypeScript code using TSDoc, it's a mistake to put types in a comment.  The initial implementation of the TSDoc parser is "strict" and reports an error.

But this practice is valid with JSDoc, since it was designed for JavaScript.  Because of that, it's somewhat common to find that notation in legacy TypeScript code.  To accommodate this, the plan is to implement a "lax" mode that allows the TSDoc parser to accept these notations.  We'd recommend to use "strict" mode for new projects, and "lax" mode for existing projects.

That's the plan, but the first priority is to finish the strict TSDoc specification.  It's close to being done, but there's a few major changes that need to get incorporated before the "1.0" release:

- Integrate the [revamped](https://github.com/microsoft/tsdoc/blob/master/tsdoc/src/beta/DeclarationReference.grammarkdown) declaration reference syntax, which was finalized a couple months ago
- Implement some missing Markdown constructs such as lists, boldface, etc.
- Expand TSDoc to include full HTML parsing, whereas currently HTML tags are treated as a flat list of tokens rather than a tree
- Make the escaping rules more rigorous, e.g. adding support for HTML entities

We're making steady progress on these things, but it takes time.

**Comment by trusktr** [comment]
> In TypeScript, types are specified using in the language itself. It doesn't make sense to declare them in the comment.

This is normally true.

The reason I want source-code-agnostic JS/TSDoc is to be very specific about my docs and how they are presented. I'd like comments to be absolute source of truth for my docs, regardless source code (though an opt-in on a case by case basis could be nice). 

Unfortunately all JSDoc tools I've tried fumble on TS code because they rely on the source (and an associated AST).

What I want to do is the following sort of documentation. Under the assumption that all classes in the project I'm documenting are mixable classes, then most files will look something like this:

<details>
<summary><code>AwesomeFeature.ts</code></summary>

```ts
// AwesomeFeature.ts

import { Mixin, MixinResult, Constructor } from 'lowclass'

// these two "mixable classes" are defined the same way as AwesomeFeature is below
import CoolFeature from './CoolFeature'
import OtherFeature from './OtherFeature'

/**
 * @class AwesomeFeature - Provides some awesome feature
 */
function AwesomeFeatureMixin<C extends Constructor>(Base: C) {
    class AwesomeFeature extends CoolFeature.mixin(OtherFeature.mixin(Constructor(Base))) {
        /**
         * @property {boolean} performingAwesome - Whether awesomeness is currently happening or not.
         */
        performingAwesome = false

        /**
         * @method doAwesome - Do awesome with this.
         * @param {number} howMuch - How much awesome to do.
         * @returns {Promise<undefined>} - Whether awesomeness succeeded or not.
         */
        async doAwesome(howMuch: number): Promise<void> {
            this.performingAwesome = true
            // ...
            this.performingAwesome = false
        }
    }

    return AwesomeFeature as MixinResult<typeof AwesomeFeature, C>
}

// export a class that can also be used as a mixin
export const AwesomeFeature = Mixin(AwesomeFeatureMixin)
export type AwesomeFeature = InstanceType<typeof AwesomeFeature>
```

The user of `AwesomeFeature` can extend it as a class, or use it as a mixin:

```ts
class Foo extends AwesomeFeature {/* ... */}
// or
class Foo extends AwesomeFeature.mixin(SomeOtherClass) {/* ... */}
```

`CoolFeature` and `OtherFeature` work the same way, as classes or mixins.

</details>

See what I'm trying to do? The mixability of my classes is inherent throughout the project, and I want to document them very simply.

Tools like TypeDoc try to document (infer) too much, and output too much noise. I don't want machine-generated "docs" for everything that can possibly be found in my source code. I want to make clean simple semantic human-readable docs, and don't mind the extra manual labor that will be needed to achieve it.

Someone interested in every single type in the code base can just open the source in VS Code.

---

I'm thinking of a possible approach: make or use a code-agnostic comment extractor, then pass that to [`Doctrine`](https://github.com/eslint/doctrine), and go from there with the Doctrine's JSON output for writing simple custom humanized docs.

**Comment by octogonz** [comment]
I see what you mean.  I'll start by saying that mixins are overcomplicated, difficult to debug, and difficult for users to understand.  Nobody should be excited about mixins.  I really mean that!

That said, now let's talk about how mixins worked awesome for my own project API Extractor.  😅  Please check out https://github.com/microsoft/tsdoc/issues/137#issuecomment-446446564 for the full details and proposal.

The *TL;DR**:
- You need to include an **interface** and/or **namespace** that tells the type system (and documentation engine) about the members of your class
- We're proposing a TSDoc tag `@partOf` that would tell the documentation engine to treat these declarations as a single entity
- We're proposing a tag `@documentAs` that tells documentation engine to display it as being a "class" or "mixin"

I didn't pursue this feature because API Extractor is literally the only project where I can justify using mixins.  But if you like this idea, we could do a quick PR to add the `@partOf` and `@documentAs` tags to TSDoc.  (And if you use API Extractor, we could try to prioritize this feature; I suspect it's relatively easy to implement.)

**Comment by trusktr** [comment]
> include an interface and/or namespace that tells the type system (and documentation engine) about the members of your class

The one in [Leafdoc](https://github.com/Leaflet/Leafdoc) is great; it supports multiple inheritance (whether that's class mixins, concatenation of objects, or whatever) that can be documented across multiple files, and is source agnostic. An `@class` comment can have any number of `@inherits` comments associated with it. However the syntax isn't exactly JSDoc, though similar.

What would a simple example of `@partOf` and `@documentAs` look like?

**Comment by octogonz** [comment]
See the other issue I linked above

**Comment by trusktr** [comment]
Ah, thanks. It seems like a little bit more complicated than I'd like.

I think of TypeScript as an implementation detail for the mixins. I'd like to document them without regards to the specific language, avoiding to document all the TypeScript-specific mechanisms that make up a mixin in TypeScript.

I really like how it works in Leafdoc:

```js
/**
 * @class Foo
 */

blah blah, this syntax { is ) not ) any & valid ] code

/**
 * This method return a number and has a string param
 * @method fooMethod: number
 * @param name: string
 */

blah blah, this syntax { is ) not ) any & valid ] code

/**
 * @class Bar
 */

blah blah, this syntax { is ) not ) any & valid ] code

/**
 * @method barMethod: void
 * @param shouldDoSomething: boolean
 */

blah blah, this syntax { is ) not ) any & valid ] code

/**
 * This class extends from both Foo and Bar
 * @class Baz
 * @inherits Foo
 * @inherits Bar
 */

blah blah, this syntax { is ) not ) any & valid ] code

/**
 * @method bazMethod: void
 */

blah blah, this syntax { is ) not ) any & valid ] code
```

This simple system allows documenting compositional patterns, without being tied to the source (`@namespace` can be used in place of `@class` to document non-class compositional patterns).

Here's the example output of the above example (my output template is WIP, it'll look better later!):

<img width="786" alt="Screen Shot 2019-08-30 at 9 33 09 PM" src="https://user-images.githubusercontent.com/297678/64059267-d2cc7800-cb6d-11e9-8b04-82d6ceb2b44f.png">

With this particular template, the inherited methods are visible in `<details>` drop downs:

<img width="456" alt="Screen Shot 2019-08-30 at 9 34 28 PM" src="https://user-images.githubusercontent.com/297678/64059276-f5f72780-cb6d-11e9-9c3f-268cabeb1abc.png">

I really like the simplicity of Leafdoc, but it would be nice to be able to achieve the same with a standard syntax like JS/TSDoc, because then text editors (f.e. VS Code) have official support for them.

**Comment by octogonz** [comment]
That makes sense. However your examples only use primitive types like `string`.

What will you do with the parameter whose type is `Map<string, Widget[]>`? How will you document generic parameter constraints? Etc

**Comment by trusktr** [comment]
> your examples only use primitive types like string.

It doesn't matter much, because it is source agnostic. I could put anything in place of `string`, like `module>foo/bar/baz.ts#FooClass`, and it just gives me that string in the output:

<img width="527" alt="Screen Shot 2019-09-01 at 12 14 47 AM" src="https://user-images.githubusercontent.com/297678/64073045-9b34fd00-cc4d-11e9-8249-bb3d1312e0f8.png">

So, the docs have any meaning I want. I can determine how to link to that module and class my own, if I want to build on top of it.

---

I decided to just make my own JSDoc parser, and then I'll do similar to Leafdoc where it'll just give me the content of the `@tag`s, as strings, then a higher-level tool could decide what to do with them.

So far, I'm [doing it with `RegExp`s](https://github.com/jonschlinkert/extract-comments/issues/19#issuecomment-526884678), but it would be nice to upgrade it to a streaming lexer later.

I imagine a higher-level tool could choose to be source-aware and to parse types, etc. F.e. [jsdoctypeparser](https://github.com/jsdoctypeparser/jsdoctypeparser).

For my case though, I'll take the strings and make something minimal (source agnostic).

---

TypeScript supports [types in the JSDoc comments](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html). I wonder what it's using for that. Maybe it's all just built into TypeScript, but it'd be nice if it could be used standalone...

**Comment by trusktr** [comment]
Maybe if there were a jsdoc-based well-known code-agnostic doc tool, things like [sveltedoc-parser](https://www.npmjs.com/package/sveltedoc-parser] or [vuedoc-parser](https://gitlab.com/vuedoc/parser) would not come to exist.

Those two particular examples are nice because they're both on top of JS, with no types in the source code anyways, so the genericness of the tool shines more with those (as opposed to documenting already-typed TS).

People using the generic doc tool could make higher level tools to generate documentation sites for different types of libs (Svelte vs Vue).

**Comment by Gerrit0** [comment]
It sounds like you are after JSDoc with the `commentsOnly` plugin. https://github.com/jsdoc/jsdoc/blob/master/plugins/commentsOnly.js

**Comment by trusktr** [comment]
Looking forward to trying the "lax" mode once that's out. :)

**Comment by trusktr** [comment]
@Gerrit0 Oh interesting, I hadn't seen that plugin before. I'll try it out...

**Comment by trusktr** [comment]
Everything I try has some critical issue. Having even a subtle format issue, like writing

```ts
/**
 * @class Foo
 * This is a cool class.
 */
```

causes JSDoc to break, and output nothing after it reaches that comment.

😫

---

It'd be great to have a tool that reads comments, and always returns the content of every comment in some understandable form without tripping up on insignificant details like above.
